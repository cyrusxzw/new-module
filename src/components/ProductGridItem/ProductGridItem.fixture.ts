const ProductGridItemFixture = {
  copy: {
    addToCart: {
      cartAction: 'copy add to cart',
      outOfStock: {
        label: 'copy out of stock label',
        title: 'copy out of stock title',
      },
      updateNotification: 'copy update notification',
    },
  },
  info: '4 Sizes / From $ 26.45',
  variantOptionsOneItem: [
    {
      size: '50 mL - In stock',
      sku: 'ARD33',
      price: '$26.45',
      isInStock: true,
      image: {
        altText: 'alt-tag-ARD33',
        sizes: {
          large: './assets/images/Product/variant-one-large.png',
          medium: './assets/images/Product/variant-one-medium.png',
          small: './assets/images/Product/variant-one-small.png',
        },
      },
    },
  ],
  variantOptions: [
    {
      size: '50 mL - In stock',
      sku: 'ARD33',
      price: '$26.45',
      isInStock: true,
      image: {
        altText: 'alt-tag-ARD33',
        sizes: {
          large: './assets/images/Product/variant-one-large.png',
          medium: './assets/images/Product/variant-one-medium.png',
          small: './assets/images/Product/variant-one-small.png',
        },
      },
    },
    {
      size: '100 mL - Not in stock',
      sku: 'ARD32',
      price: '$86.97',
      isInStock: false,
      image: {
        altText: 'alt-tag-ARD32',
        sizes: {
          large: './assets/images/Product/variant-two-large.png',
          medium: './assets/images/Product/variant-two-medium.png',
          small: './assets/images/Product/variant-two-small.png',
        },
      },
    },
    {
      size: '180 mL - In stock',
      sku: 'ARD30',
      price: '$96.97',
      isInStock: true,
      image: {
        altText: 'alt-tag-ARD32',
        sizes: {
          large: './assets/images/Product/variant-two-large.png',
          medium: './assets/images/Product/variant-two-medium.png',
          small: './assets/images/Product/variant-two-small.png',
        },
      },
    },
  ],
  url: '/p/',
};

export { ProductGridItemFixture };
