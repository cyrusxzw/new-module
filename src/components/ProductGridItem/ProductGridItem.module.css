@import 'styles/index';

.base {
  width: 100%;
  transition: background-color 300ms var(--easing-ease-out-cubic);

  @media (--viewport-xs-sm-only) {
    padding: var(--layout-lg-spacing) 20px;
  }

  &,
  & * {
    box-sizing: border-box;
  }

  @media (--viewport-md-only) {
    padding: var(--layout-md-spacing);
  }

  @media (--viewport-lg) {
    position: relative;
    padding: var(--layout-lg-spacing) var(--layout-lg-spacing) 65px;

    &:hover {
      background-color: var(--color-white-darker-2);

      .variants,
      .addToCartButton {
        opacity: 1;
      }

      .infoHolder:not(.hasOneVariant) {
        opacity: 0;
      }
    }
  }
}

.image {
  display: flex;
  align-items: flex-end;
  justify-content: center;

  @media (--viewport-xs-sm-only) {
    padding-top: 50px;
  }
  @media (--viewport-md-only) {
    height: 300px;
  }

  @media (--viewport-lg) {
    height: 490px;
  }

  img {
    width: auto;
    max-width: 100%;
    flex-shrink: 0; /* IE */

    @media (--viewport-xs-sm-only) {
      max-height: 220px;
    }
    @media (--viewport-md) {
      width: 180px;
      max-height: 100%;
    }
    @media (--viewport-lg) {
      width: 330px;
    }
  }
}

.imageLink {
  display: block;
}

.productName {
  margin-top: 0;
  margin-bottom: 7px;
  text-align: center;

  a {
    font-family: inherit;
  }
}

.productNameLink {
  color: currentColor;
}

.infoVariantHolder {
  @media (--viewport-xs-sm-md-only) {
    margin-bottom: 24px;
  }
  @media (--viewport-md) {
    min-height: 60px;
  }
  @media (--viewport-lg) {
    &:focus-within .infoHolder {
      opacity: 0;
    }
  }
}

.infoHolder {
  @media (--viewport-xs-sm-md-only) {
    opacity: 1;
  }
  @media (--viewport-lg) {
    transition: opacity 300ms var(--easing-ease-out-cubic);
  }

  &.hasManyVariants {
    @media (--viewport-xs-sm-md-only) {
      display: none;
    }
  }
}

.info {
  margin: -2px 0; /* to match radio button negative margins */
  color: var(--color-grey-30);
  font-size: rem(14px);
  line-height: 1.6;
  text-align: center;
}

.variants {
  @media (--viewport-lg) {
    position: relative;
    z-index: 2;
    margin-top: -26px;
    opacity: 0;
    transition: opacity 300ms var(--easing-ease-out-cubic);

    &:focus-within {
      opacity: 1;
    }
  }
}

.addToCartButton {
  width: 100%;
  max-width: 100%;
  text-align: center;

  @media (--viewport-lg) {
    position: absolute;
    right: 0;
    bottom: 0;
    left: 0;
    max-width: 100%;
    opacity: 0;

    &:focus {
      opacity: 1;
    }
  }
}

.ctaLink {
  display: block;
  height: 60px;
  padding: 20px 15px;
  border-width: 1px;
  border-style: solid;
  border-color: var(--color-dark-background);
  background-color: var(--color-dark-background);
  color: var(--color-light-link-hover);
  font-family: var(--font-family-sans-serif-medium);
  font-size: rem(14px);
  transition: all 200ms var(--easing-ease-in-out-cubic);

  &:hover,
  &:focus {
    border-color: var(--color-black);
    background-color: var(--color-black);
    color: var(--color-light-link-hover);
  }
}

.definitionList {
  display: grid;
  grid-template-columns: auto minmax(0, 1fr);

  @media (--viewport-xs-sm-md-only) {
    margin-top: 30px;
  }

  @media (--viewport-lg) {
    .dark & {
      border-top: 2px solid var(--color-grey-20);
    }

    .light & {
      border-top: 2px solid var(--color-light-copy);
    }
  }

  dt {
    padding-right: 20px;
    border-bottom: 1px solid var(--color-grey-65);
    margin-bottom: 0;
  }

  dt,
  dd {
    padding-top: 20px;
    padding-bottom: 20px;
  }

  dt:nth-of-type(1) {
    @media (--viewport-lg) {
      padding-top: 20px;
    }
  }

  dd:nth-of-type(1) {
    @media (--viewport-xs-sm-md-only) {
      padding-top: 0;
    }
  }

  @media all and (-ms-high-contrast: none) {
    display: flex;
    flex-wrap: wrap;

    dt {
      max-width: 35%;
      flex: 0 0 35%;
    }

    dd {
      max-width: 65%;
      flex: 0 0 65%;
    }
  }
}

/* Overwrites for smaller desktop overlapping into tablet */
@media screen and (min-width: 1000px) and (hover: hover) {
  .addToCartButton {
    position: absolute;
    right: 0;
    bottom: 0;
    left: 0;
    max-width: 100%;
    opacity: 0;

    &:focus {
      opacity: 1;
    }
  }

  .variants {
    position: relative;
    z-index: 2;
    margin-top: -26px;
    opacity: 0;
    transition: opacity 300ms var(--easing-ease-out-cubic);

    &:focus-within {
      opacity: 1;
    }
  }

  .base {
    position: relative;
    padding: var(--layout-lg-spacing) var(--layout-lg-spacing) 65px;

    &:hover {
      background-color: var(--color-white-darker-2);

      .variants,
      .addToCartButton {
        opacity: 1;
      }

      .infoHolder:not(.hasOneVariant) {
        opacity: 0;
      }
    }
  }

  .infoHolder.hasManyVariants.hasManyVariants {
    display: block;
  }

  .image {
    height: 490px;

    img {
      width: 330px;
    }
  }

  .definitionList {
    .dark & {
      border-top: 2px solid var(--color-grey-20);
    }

    .light & {
      border-top: 2px solid var(--color-light-copy);
    }

    dt:nth-of-type(1),
    dd:nth-of-type(1):nth-of-type(1) {
      padding-top: 20px;
    }
  }
}
