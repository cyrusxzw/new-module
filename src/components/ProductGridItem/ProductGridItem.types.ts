import type { Themes } from '~/types';

type Copy = {
  addToCart?: {
    cartAction?: string;
    outOfStock?: {
      label?: string;
      title?: string;
    };
    updateNotification?: string;
  };
};

type CallToAction = {
  clickFunction?: () => void;
  text?: string;
  url?: string;
};

type ProductGridItemProps = {
  className?: string;
  copy?: Copy;
  cta?: CallToAction;
  id?: string;
  info?: string;
  onCtaClick?: () => void;
  theme?: Themes;
  url?: string;
};

export type { ProductGridItemProps };
