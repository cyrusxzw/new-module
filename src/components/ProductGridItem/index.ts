import { ProductGridItem } from './ProductGridItem';

export { ProductGridItem };
