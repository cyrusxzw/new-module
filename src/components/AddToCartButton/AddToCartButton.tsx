import React from 'react';
import cx from 'classnames';
import { useAddToCartContext, useVariantSelectContext } from '~/contexts';
import { HYPERLINK_STYLE_TYPES } from '~/constants';
import { Button } from '~/components/Button';
import { Loading } from '~/components/Loading';
import { Hyperlink } from '~/components/Hyperlink';
import { useDelayedDisabled } from './AddToCartButton.utils';
import type { AddToCartButtonType } from './AddToCartButton.types';
import styles from './AddToCartButton.module.css';

const AddToCartButton: AddToCartButtonType = ({
  className,
  copy,
  dataTestRef = 'ADD_TO_CART',
  isEnabled = true,
  isFullWidth = true,
  theme = 'dark',
}) => {
  const addToCartContext = useAddToCartContext();
  const { selectedVariant } = useVariantSelectContext();
  const {
    errorMessage,
    hasError,
    isLoading,
    isUpdateSuccessful,
  } = addToCartContext;
  const shouldShowUpdateSuccessMessage = !isLoading && isUpdateSuccessful;
  const isDelayedDisabled = useDelayedDisabled(shouldShowUpdateSuccessMessage);
  const updateNotificationLabel = copy?.updateNotification;

  if (!selectedVariant) return null;

  const {
    alternateAction,
    isInStock,
    isSellable,
    price,
    sku,
  } = selectedVariant;

  const cartActionText = copy?.cartAction
    ? `${copy?.cartAction} â€” ${price}`
    : price;
  const postTaxPriceText = copy?.postTaxPrice ? ` ${copy?.postTaxPrice}` : '';
  const cartActionLabel = cartActionText + postTaxPriceText;

  const classSet = cx(
    styles.base,
    { [styles.fullWidth]: isFullWidth },
    className,
  );

  const handleOnClick = () => {
    const { dispatch, onClick } = addToCartContext;

    onClick(sku, dispatch);
  };

  if (!!alternateAction?.url && !!alternateAction?.label) {
    return (
      <Hyperlink
        className={classSet}
        dataTestRef={dataTestRef}
        hasTargetInNewWindow={alternateAction?.openInANewWindow}
        isAlternate={true}
        style={HYPERLINK_STYLE_TYPES.INTERNAL_NO_ICON_BUTTON_LINK}
        theme={theme}
        title={alternateAction?.label}
        url={alternateAction?.url}
      >
        {alternateAction?.label}
      </Hyperlink>
    );
  }

  if (hasError) {
    /** @TODO Handle errors thrown by handleOnClick */
    console.error('Add To Cart button updateError: ', errorMessage); // eslint-disable-line
  }

  const isButtonEnabled =
    !isLoading &&
    !!price &&
    !!sku &&
    isInStock &&
    !!sku &&
    isEnabled &&
    !hasError &&
    !isDelayedDisabled &&
    isSellable;

  const labelClassName = cx(
    styles.label,
    { [styles.hideLabel]: isLoading },
    { [styles.showSuccessMessage]: shouldShowUpdateSuccessMessage },
  );

  return (
    <Button
      className={classSet}
      dataTestRef={dataTestRef}
      isAlternate={true}
      isEnabled={isButtonEnabled}
      onClick={handleOnClick}
      theme={theme}
      title={isInStock ? cartActionLabel : copy?.outOfStock?.title}
    >
      {isLoading && (
        <Loading
          className={styles.loading}
          data-test-ref={`${dataTestRef}_UPDATING`}
          isLoading={true}
          theme={theme === 'dark' ? 'light' : 'dark'}
        />
      )}
      <span
        className={labelClassName}
        data-test-ref={
          shouldShowUpdateSuccessMessage ? `${dataTestRef}_SUCCESS` : undefined
        }
      >
        {shouldShowUpdateSuccessMessage ? (
          <span>{updateNotificationLabel}</span>
        ) : (
          <span> </span>
        )}
        <span>{isInStock ? cartActionLabel : copy?.outOfStock?.label}</span>
      </span>
    </Button>
  );
};

export { AddToCartButton };
