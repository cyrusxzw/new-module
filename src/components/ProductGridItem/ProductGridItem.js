import React, { useRef } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { HEADING } from '~/constants';
import { useProductDetailContext, useVariantSelectContext } from '~/contexts';
import { useImageTransition } from '~/customHooks';
import { getVariantRadioOptions } from '~/utils/product/index.ts';
import { AddToCartButton } from '~/components/AddToCartButton';
import { DefinitionList } from '~/components/DefinitionList/index.ts';
import { Heading } from '~/components/Heading/index.ts';
import { Hyperlink } from '~/components/Hyperlink';
import { Image } from '~/components/Image';
import { P } from '~/components/Paragraph';
import { RadioGroup } from '~/components/RadioGroup/index.ts';
import { Transition } from '~/components/Transition';
import styles from './ProductGridItem.module.css';

const ProductGridItem = React.forwardRef(
  ({ className, onCtaClick, copy, cta, id, info, theme, url }, ref) => {
    const imageRef = useRef();
    const {
      selectedVariant,
      onVariantChange,
      variants,
    } = useVariantSelectContext();
    const { productDetail } = useProductDetailContext();

    const hasOneVariant = variants?.length === 1;
    const isBundle = variants?.length === 0;

    const [currentImage, isImageActive] = useImageTransition(
      selectedVariant?.image || productDetail.image,
      imageRef,
    );

    if (!productDetail) return null;

    const { definitionList, productName, sku } = productDetail;

    const variantRadioOptions = getVariantRadioOptions(variants);
    const classSet = cx(styles.base, styles[theme], className);

    const classInfoHolderSet = cx(styles.infoHolder, {
      [styles.hasOneVariant]: hasOneVariant || isBundle,
      [styles.hasManyVariants]: !hasOneVariant,
    });

    const classCtaLinkSet = cx(styles.addToCartButton, styles.ctaLink);

    const RADIO_GROUP_NAME = sku;
    const RADIO_GROUP_DATA_TEST_REF = 'PRODUCT_GRID_ITEM_VARIANT_SELECT';

    return (
      <div className={classSet} id={id} ref={ref}>
        <Transition isActive={isImageActive} name="fade">
          <Hyperlink
            className={styles.imageLink}
            onClick={() => onCtaClick()}
            url={url}
          >
            <Image
              altText={currentImage.altText}
              className={styles.image}
              ref={imageRef}
              sizes={currentImage.sizes}
            />
          </Hyperlink>
        </Transition>

        <Heading
          className={styles.productName}
          hasMediumWeightFont={true}
          level={HEADING.LEVEL.FIVE}
          size={HEADING.SIZE.X_X_SMALL}
          theme={theme}
        >
          <Hyperlink className={styles.productNameLink} url={url}>
            {productName}
          </Hyperlink>
        </Heading>

        <div className={styles.infoVariantHolder}>
          <div className={classInfoHolderSet}>
            <P className={styles.info} theme={theme}>
              {info}
            </P>
          </div>

          {!hasOneVariant && !isBundle && (
            <RadioGroup
              className={styles.variants}
              dataTestRef={RADIO_GROUP_DATA_TEST_REF}
              name={RADIO_GROUP_NAME}
              onChange={e => onVariantChange(e, variants)}
              options={variantRadioOptions}
              theme={theme}
              value={selectedVariant.sku}
            />
          )}
        </div>

        {!isBundle && (
          <AddToCartButton
            className={styles.addToCartButton}
            copy={copy.addToCart}
            isFullWidth={true}
            theme={theme}
          />
        )}

        {cta && isBundle && (
          <Hyperlink
            className={classCtaLinkSet}
            isAlternate={false}
            onClick={() => onCtaClick()}
            theme={theme}
            title={cta?.text}
            url={cta?.url}
          >
            {cta?.text}
          </Hyperlink>
        )}

        <DefinitionList
          className={styles.definitionList}
          hasBottomBorder={true}
          items={definitionList}
          theme={theme}
        />
      </div>
    );
  },
);

ProductGridItem.propTypes = {
  className: PropTypes.string,
  onCtaClick: PropTypes.func,
  copy: PropTypes.shape({
    addToCart: PropTypes.shape({
      cartAction: PropTypes.string,
      updateNotification: PropTypes.string,
      outOfStock: PropTypes.shape({
        label: PropTypes.string,
        title: PropTypes.string,
      }),
    }),
  }),
  cta: PropTypes.shape({
    text: PropTypes.string,
    url: PropTypes.string,
    clickFunction: PropTypes.func,
  }),
  id: PropTypes.string,
  info: PropTypes.string,
  theme: PropTypes.oneOf(['dark', 'light']),
  url: PropTypes.string,
};

ProductGridItem.defaultProps = {
  className: undefined,
  copy: {
    addToCart: {
      cartAction: undefined,
      updateNotification: undefined,
      outOfStock: {
        label: undefined,
        title: undefined,
      },
    },
  },
  cta: undefined,
  id: undefined,
  info: undefined,
  onCtaClick: () => undefined,
  theme: 'dark',
  url: undefined,
};

export { ProductGridItem };
