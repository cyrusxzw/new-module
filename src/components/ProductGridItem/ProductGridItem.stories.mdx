import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import * as knobs from '@storybook/addon-knobs';
import { StorybookGridOverlay } from '~.storybook/components';
import {
  AddToCartContextProvider,
  ProductDetailContextProvider,
  VariantSelectContextProvider,
} from '~/contexts';
import { ProductDetailHeaderFixture } from '~/components/ProductDetailHeader/ProductDetailHeader.fixture';
import { AddToCartButtonFixture } from '~/components/AddToCartButton/AddToCartButton.fixture';
import { mockAddToCartButtonOnClick } from '~/components/AddToCartButton/__mocks__/AddToCartButton.onClick';
import { ProductGridItem } from './ProductGridItem';
import { ProductGridItemFixture } from './ProductGridItem.fixture';

<Meta title="Components/ProductGridItem" component={ProductGridItem} />

# ProductGridItem

<Canvas>
  <Story name="Base component">
    {(() => {
      const theme = knobs.select(
        'theme',
        ['dark', 'light'],
        'dark',
        'Presentation',
      );
      const implementation = {
        gridOverlay: {
          isActive: knobs.boolean('Show grid overlay', false, 'Implementation'),
          hasInvertedColours: knobs.boolean(
            'Invert grid overlay colours',
            false,
            'Implementation',
          ),
        },
      };
      return (
        <>
        <AddToCartContextProvider onClick={mockAddToCartButtonOnClick}>
          <ProductDetailContextProvider
            product={knobs.object('Product', ProductDetailHeaderFixture.product, 'Content')}
          >
            <VariantSelectContextProvider
              variants={knobs.object('Variant', ProductGridItemFixture.variantOptions, 'Content')}
            >
              <ProductGridItem
                copy={{
                  addToCart: AddToCartButtonFixture.copy,
                  ...ProductGridItemFixture.copy,
                }}
                url={knobs.text('URL', ProductGridItemFixture.url, 'Content')}
                theme={theme}
              />
            </VariantSelectContextProvider>
          </ProductDetailContextProvider>
        </AddToCartContextProvider>
          <StorybookGridOverlay
            isActive={implementation.gridOverlay.isActive}
            hasInvertedColours={implementation.gridOverlay.hasInvertedColours}
          />
        </>
      );
    })()}
  </Story>
</Canvas>

## API

<ArgsTable of={ProductGridItem} />
