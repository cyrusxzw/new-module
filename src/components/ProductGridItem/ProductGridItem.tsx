import React, { useRef, forwardRef } from 'react';
import cx from 'classnames';
import {
  useThemeContext,
  useProductDetailContext,
  useVariantSelectContext,
} from '~/contexts';
import { useImageTransition } from '~/customHooks';
import { getVariantRadioOptions } from '~/utils/product/index.ts';
import { AddToCartButton } from '~/components/AddToCartButton';
import { DefinitionList } from '~/components/DefinitionList/index.ts';
import { Heading } from '~/components/Heading/index.ts';
import { Hyperlink } from '~/components/Hyperlink';
import { Image } from '~/components/Image';
import { P } from '~/components/Paragraph';
import { RadioGroup } from '~/components/RadioGroup/index.ts';
import { Transition } from '~/components/Transition';
import type { ProductGridItemProps } from './ProductGridItem.types';
import styles from './ProductGridItem.module.css';

const ProductGridItem = forwardRef<HTMLDivElement, ProductGridItemProps>(
  function ProductGridItemRef(
    { className, onCtaClick, copy, cta, id, info, theme, url },
    ref,
  ) {
    const currentTheme = useThemeContext(theme, 'dark');
    const imageRef = useRef();
    const { productDetail } = useProductDetailContext();

    const {
      selectedVariant,
      onVariantChange,
      variants,
    } = useVariantSelectContext();

    const [currentImage, isImageActive] = useImageTransition(
      selectedVariant?.image || productDetail.image,
      imageRef,
    );

    if (!productDetail) return null;

    const hasOneVariant = variants?.length === 1;
    const isBundle = variants?.length === 0;
    const { definitionList, productName, sku } = productDetail;
    const variantRadioOptions = getVariantRadioOptions(variants);
    const handleOnCtaClick = () => onCtaClick?.();

    const classSet = cx(styles.base, styles[currentTheme], className);

    const classInfoHolderSet = cx(styles.infoHolder, {
      [styles.hasOneVariant]: hasOneVariant || isBundle,
      [styles.hasManyVariants]: !hasOneVariant,
    });

    const classCtaLinkSet = cx(styles.addToCartButton, styles.ctaLink);

    const RADIO_GROUP_NAME = sku;
    const RADIO_GROUP_DATA_TEST_REF = 'PRODUCT_GRID_ITEM_VARIANT_SELECT';

    return (
      <div className={classSet} id={id} ref={ref}>
        <Transition isActive={isImageActive} type="fade">
          <Hyperlink
            className={styles.imageLink}
            onClick={handleOnCtaClick}
            url={url}
          >
            <Image
              altText={currentImage.altText}
              className={styles.image}
              ref={imageRef}
              sizes={currentImage.sizes}
            />
          </Hyperlink>
        </Transition>

        <Heading
          className={styles.productName}
          hasMediumWeightFont={true}
          level="5"
          size="xXSmall"
          theme={theme}
        >
          <Hyperlink className={styles.productNameLink} url={url}>
            {productName}
          </Hyperlink>
        </Heading>

        <div className={styles.infoVariantHolder}>
          <div className={classInfoHolderSet}>
            <P className={styles.info} theme={theme}>
              {info}
            </P>
          </div>

          {!hasOneVariant && !isBundle && (
            <RadioGroup
              className={styles.variants}
              dataTestRef={RADIO_GROUP_DATA_TEST_REF}
              name={RADIO_GROUP_NAME}
              onChange={(event) => onVariantChange(event, variants)}
              options={variantRadioOptions}
              theme={theme}
              value={selectedVariant?.sku}
            />
          )}
        </div>

        {!isBundle && (
          <AddToCartButton
            className={styles.addToCartButton}
            copy={copy.addToCart}
            dataTestRef="PRODUCT_GRID_ITEM"
            isFullWidth={true}
            theme={theme}
          />
        )}

        {cta && isBundle && (
          <Hyperlink
            className={classCtaLinkSet}
            isAlternate={false}
            onClick={handleOnCtaClick}
            theme={theme}
            title={cta?.text}
            url={cta?.url}
          >
            {cta?.text}
          </Hyperlink>
        )}

        <DefinitionList
          className={styles.definitionList}
          hasBottomBorder={true}
          items={definitionList}
          theme={theme}
        />
      </div>
    );
  },
);

export { ProductGridItem };
