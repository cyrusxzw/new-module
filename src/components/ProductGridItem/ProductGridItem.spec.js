import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import {
  AddToCartContextProvider,
  ProductDetailContextProvider,
  VariantSelectContextProvider,
} from '~/contexts';
import { ProductDetailHeaderFixture } from '~/components/ProductDetailHeader/ProductDetailHeader.fixture';
import { AddToCartButtonFixture } from '~/components/AddToCartButton/AddToCartButton.fixture';
import { mockAddToCartButtonOnClick } from '~/components/AddToCartButton/__mocks__/AddToCartButton.onClick';
import { ProductGridItem } from './ProductGridItem';
import { ProductGridItemFixture } from './ProductGridItem.fixture';

describe('<ProductGridItem />', () => {
  it('should render base component correctly', () => {
    const { container } = render(
      <AddToCartContextProvider onClick={mockAddToCartButtonOnClick}>
        <ProductDetailContextProvider
          product={ProductDetailHeaderFixture.product}
        >
          <VariantSelectContextProvider
            variants={ProductGridItemFixture.variantOptions}
          >
            <ProductGridItem
              copy={{
                addToCart: AddToCartButtonFixture.copy,
                ...ProductGridItemFixture.copy,
              }}
              info={ProductGridItemFixture.info}
              url={ProductGridItemFixture.url}
            />
          </VariantSelectContextProvider>
        </ProductDetailContextProvider>
      </AddToCartContextProvider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render with isBundle being true', () => {
    const bundleLinkName = "I'll give you an offer you can't refuse";
    const onCtaClickMock = jest.fn();

    render(
      <AddToCartContextProvider onClick={mockAddToCartButtonOnClick}>
        <ProductDetailContextProvider
          product={ProductDetailHeaderFixture.product}
        >
          <VariantSelectContextProvider variants={[]}>
            <ProductGridItem
              copy={{
                addToCart: AddToCartButtonFixture.copy,
                ...ProductGridItemFixture.copy,
              }}
              cta={{ text: bundleLinkName, url: 'corleone' }}
              info={ProductGridItemFixture.info}
              onCtaClick={onCtaClickMock}
              url={ProductGridItemFixture.url}
            />
          </VariantSelectContextProvider>
        </ProductDetailContextProvider>
      </AddToCartContextProvider>,
    );
    const link = screen.getByRole('link', { name: bundleLinkName });
    userEvent.click(link);

    expect(onCtaClickMock).toHaveBeenCalledTimes(1);
  });

  it('should execute default props version of cta click correctly', () => {
    render(
      <AddToCartContextProvider onClick={mockAddToCartButtonOnClick}>
        <ProductDetailContextProvider
          product={ProductDetailHeaderFixture.product}
        >
          <VariantSelectContextProvider
            variants={ProductGridItemFixture.variantOptions}
          >
            <ProductGridItem
              copy={{
                addToCart: AddToCartButtonFixture.copy,
                ...ProductGridItemFixture.copy,
              }}
              info={ProductGridItemFixture.info}
              url={ProductGridItemFixture.url}
            />
          </VariantSelectContextProvider>
        </ProductDetailContextProvider>
      </AddToCartContextProvider>,
    );

    const link = screen.getAllByRole('link')[0]; // First Hyperlink with click function
    userEvent.click(link);
    // TODO: Assert default function situation
  });

  it('should execute cta click correctly', () => {
    const onCtaClickMock = jest.fn();
    render(
      <AddToCartContextProvider onClick={mockAddToCartButtonOnClick}>
        <ProductDetailContextProvider
          product={ProductDetailHeaderFixture.product}
        >
          <VariantSelectContextProvider
            variants={ProductGridItemFixture.variantOptions}
          >
            <ProductGridItem
              copy={{
                addToCart: AddToCartButtonFixture.copy,
                ...ProductGridItemFixture.copy,
              }}
              info={ProductGridItemFixture.info}
              onCtaClick={onCtaClickMock}
              url={ProductGridItemFixture.url}
            />
          </VariantSelectContextProvider>
        </ProductDetailContextProvider>
      </AddToCartContextProvider>,
    );

    const link = screen.getAllByRole('link')[0]; // First Hyperlink with click function
    userEvent.click(link);

    expect(onCtaClickMock).toHaveBeenCalledTimes(1);
  });
});
