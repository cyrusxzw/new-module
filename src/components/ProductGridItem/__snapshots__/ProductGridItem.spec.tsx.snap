// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductGridItem /> should render base component correctly 1`] = `
<div>
  <div
    class="base dark"
  >
    <a
      class="base left dark imageLink fadeEnter"
      href="/p/"
      target="_self"
    >
      <picture
        class="base picture dark image"
      >
        <img
          loading="lazy"
        />
      </picture>
    </a>
    <h5
      class="base mediumWeightFont xXSmall dark productName"
    >
      <a
        class="base left dark productNameLink"
        href="/p/"
        target="_self"
      >
        Lorem ipsum dolor
      </a>
    </h5>
    <div
      class="infoVariantHolder"
    >
      <div
        class="infoHolder hasManyVariants"
      >
        <p
          class="base dark info"
        >
          4 Sizes / From $ 26.45
        </p>
      </div>
      <ul
        class="base variants"
      >
        <li
          class="radio"
        >
          <label
            class="label"
            for="option-ARD33"
          >
            <input
              aria-checked="true"
              class="input"
              data-test-ref="PRODUCT_GRID_ITEM_VARIANT_SELECT"
              id="option-ARD33"
              name="ARD33"
              tabindex="0"
              type="radio"
              value="ARD33"
            />
            <span
              class="pot dark"
            />
            <span
              class="labelContent dark"
            >
              50 mL - In stock
            </span>
          </label>
        </li>
        <li
          class="radio"
        >
          <label
            class="label"
            for="option-ARD32"
          >
            <input
              aria-checked="false"
              class="input"
              data-test-ref="PRODUCT_GRID_ITEM_VARIANT_SELECT"
              id="option-ARD32"
              name="ARD33"
              tabindex="0"
              type="radio"
              value="ARD32"
            />
            <span
              class="pot dark"
            />
            <span
              class="labelContent dark"
            >
              100 mL - Not in stock
            </span>
          </label>
        </li>
        <li
          class="radio"
        >
          <label
            class="label"
            for="option-ARD30"
          >
            <input
              aria-checked="false"
              class="input"
              data-test-ref="PRODUCT_GRID_ITEM_VARIANT_SELECT"
              id="option-ARD30"
              name="ARD33"
              tabindex="0"
              type="radio"
              value="ARD30"
            />
            <span
              class="pot dark"
            />
            <span
              class="labelContent dark"
            >
              180 mL - In stock
            </span>
          </label>
        </li>
      </ul>
    </div>
    <button
      class="base alternate blockStyle dark base fullWidth addToCartButton"
      data-test-ref="PRODUCT_GRID_ITEM"
      title="copy add to cart — $26.45"
      type="button"
    >
      <span
        class="label"
      >
        <span>
           
        </span>
        <span>
          copy add to cart — $26.45
        </span>
      </span>
    </button>
    <dl
      class="base dark hasBottomBorder definitionList"
      data-testid="data-testid-DefinitionList"
    >
      <dt
        class="term slideIn"
      >
        Lorem ipsum dolor
      </dt>
      <dd
        class="description slideIn"
      >
        Lorem ipsum dolort, consectetur adipiscing elit.
      </dd>
      <dt
        class="term slideIn"
      >
        Ut consectetur mi
      </dt>
      <dd
        class="description slideIn"
      >
        Ut consectetur, vitae libero imperdiet id.
      </dd>
      <dt
        class="term slideIn"
      >
        Pellentesque tristique
      </dt>
      <dd
        class="description slideIn"
      >
        Maecenas mattis, urna vel commodo vehicula
      </dd>
    </dl>
  </div>
</div>
`;
